### PR TITLE
KeepAlive Test cases and support for Http 1.0 in Client

### PR DESCRIPTION
**Pull request by sumawa**
* KeepAlive Test cases
* Based on discussion  #592 
* Improving coverage, more Server configuration tests will be added.

---

There are two standard ways of verifying KeepAlive feature:

### Keep Alive Test case 1

https://serverfault.com/a/554668

Getting "left intact" message while using curl

	* Getting last two lines indicating keepAlive is still enabled 
```
		* Connection #0 to host localhost left intact
		* Closing connection 0
```
	* And if keep-alive is not working, then it just outputs this line:
```
		* Closing connection #0
```
### Keep Alive Test case 2

As per RFC  https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.10
```
HTTP/1.1 defines the "close" connection option for the sender to signal that the connection will be closed after completion of the response. For example,

       Connection: close

in either the request or the response header fields indicates that the connection SHOULD NOT be considered `persistent' (section 8.1) after the current request/response is complete.
HTTP/1.1 applications that do not support persistent connections MUST include the "close" connection option in every message.
```

### Current behavior in zio-Http

We intend to enable / disable "keep alive" using 
```scala
  private val server =
    Server.port(PORT) ++              // Setup port
      Server.paranoidLeakDetection ++ // Paranoid leak detection (affects performance)
      Server.app(fooBar +++ app)   //++     // Setup the Http app
//      Server.keepAlive.     // uncomment to enable keep alive 

```

##### For Http 1.0
```
curl  --http1.0 -Iv http://localhost:8090/foo  
```
We get connection: close in response header (correct behavior) indicating no re-use
```
	< connection: close
	* Closing connection 0	
```

##### For Http 1.1 (Without Connection: close in request header)

We get "left intact" message indicating  re-use (correct behaviour)
```
curl  --http1.1 -Iv http://localhost:8090/foo
	< 
	* Connection #0 to host localhost left intact
	* Closing connection 0
```

##### For Http 1.1 (WITH Connection: close in request header)
```
curl  --http1.1 -Iv -H "Connection: Close" http://localhost:8090/foo
```
We get Connection: close in response header indicating no re-use (correct behaviour)
```
	< connection: close
	* Closing connection 0
```
#### Scenario 2. Disabled Server.keepAlive while creating Server

The "proof" that keepAlive is still working even if we don't use Server.keepAlive 

Run HelloWorldAdvanced in examples test (with Server.keepAlive disabled), and fire these curl requests from command line 
```
 curl -Iv http://localhost:8090/foo --next http://localhost:8090/foo 2>&1 | grep -i '#0'
```
We see the message regarding Re-use 
```
* Connected to localhost (::1) port 8090 (#0)
* Connection #0 to host localhost left intact
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (::1) port 8090 (#0)
* Connection #0 to host localhost left intact	
```
### Recommendation 
	* Allow keepAlive always and remove Server.keepAlive
	* The user requests can use "Connection: close" in the request header to control re-use mechanism
	* Also may be we can explore keepAliveTimeout mechanism rather, if there is a substantial use case.

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433742.48**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436303.63**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 437446.48**

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 1]
```diff
@@ -40,10 +50,17 @@ abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
...
     val data = HttpData.fromText(content)
     Client.request(
-      Client.ClientParams(method -> URL(path, Location.Absolute(Scheme.HTTP, "localhost", port)), headers, data),
+      Client.ClientParams(
+        method -> URL(path, Location.Absolute(Scheme.HTTP, "localhost", port)),
+        headers,
+        data,
+        null,
+        httpVersion,
+      ),
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]:
```suggestion
      Client.ClientParams(
        endpoint = method -> URL(path, Location.Absolute(Scheme.HTTP, "localhost", port)),
        getHeaders = headers,
        data = data,
        httpVersion= httpVersion,
      ),
```

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          assertM(res)(isSome(equalTo("close")))
+        } +
+        testM("For Http 1.0 any request SHOULD respond with 'connection: close' response header indicating no re-use") {
+          val path    = !!
+          val headers = List()
+          val res     =
+            app1.request(path, Method.GET, "", headers, HttpVersion.HTTP_1_0).map(_.getHeaderValue("Connection"))
+          assertM(res)(isSome(equalTo("close")))
+        }
+    } // +
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: delete commented code

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433169.16**

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    From github discussion (Keep Alive Test case 2, Scenario 1)
+
+      https://github.com/dream11/zio-http/discussions/592
+   */
+  def keepAliveSpec = suite("KeepAlive Test cases") {
+    suite("Connection: close request header test with Server KeepAlive ENABLED") {
+      val connectionCloseHeader = Header(HttpHeaderNames.CONNECTION.toString, HttpHeaderValues.CLOSE.toString)
+      val app1                  = Http.empty
+      testM(
+        "Http 1.1 WITHOUT 'Connection: close' in the request header SHOULD respond WITHOUT 'Connection: close' in the response header, indicating re-use",
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: Tests should have a concise description.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 436130.98**

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala (hunk 3 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: Made description much shorter.

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala (hunk 1 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: Done.

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala (hunk 2 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: Done

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import zio.test.Assertion.{equalTo, isNone, isSome}
+import zio.test.assertM
+
+object ServerConfigSpec extends HttpRunnableSpec(8088) {
+
+  /*
+    From github discussion (Keep Alive Test case 2, Scenario 1)
+
+      https://github.com/dream11/zio-http/discussions/592
+   */
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: follow standard scala doc settings.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440594.74**

File: zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala (hunk 4 above)
Comment (sumawa) [reply zio-http/src/test/scala/zhttp/service/ServerConfigSpec.scala]: Removed comments.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 446206.56**

**Pull request closed without merging**